### PR TITLE
ENH: Add a Bunch-like class to use as a backwards compatible return value.

### LINKED ISSUE #3665: Return Bunch objects from stats functions
Some functions like `linregress` and `theilslopes` return tuples of 4 or 5 results. A `Bunch` pattern would be much cleaner, and also allows adding extra return values easily.

See discussion at https://github.com/rgommers/scipy/commit/6c1048f6416 with @josef-pkt

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
The new function `make_tuple_bunch()` is intended to replace the use
of `namedtuple` when we want to add additional outputs to a function
while maintaining backwards compatibility.

Closes gh-3665.

### CODE REVIEW

**Comment by larsoner** [comment]
I guess one important consideration / potential drawback is that this is backward incompatible in the sense that `isinistance(output, tuple)` no longer passes. I suppose that's okay since doing `tuple(output)` should get you back to the old behavior for cases where actually getting a built-in return type is important.

File: scipy/_lib/_bunch.py
Comment (larsoner) [inline scipy/_lib/_bunch.py:41]: If you also add a `__iter__` method like:
```
    def __iter__(self):
        for ii in range(len(self)):
            yield self[ii]
```
You can inherit from `collections.abc.Sequence`:

https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes

Then you get some mixin methods for free, and people can iterate over the output (which they might have done previously) with the tuple.

Comment (WarrenWeckesser) [reply scipy/_lib/_bunch.py:41]: While I'm not saying we shouldn't do that, it looks like it isn't necessary--you can already iterate over the output:
```
In [7]: Result = make_tuple_bunch('Result', ['x', 'y'], ['w', 'beta'])

In [8]: result = Result(x=1, y=2, w=99, beta=0.5)

In [9]: for val in result:
   ...:     print(val) 
   ...:                                                                         
1
2
```

**Comment by WarrenWeckesser** [comment]
> ... `isinistance(output, tuple)` no longer passes.

Right, and I didn't find a way to make the new output pass that test and still be picklable.  (But I'm not Raymond Hettinger--maybe my Python wizarding skills just aren't strong enough.)  The new object also doesn't have the `count()` or `index()` methods that a `Sequence` would have.  But do we really need those?  I find it hard to imagine a case where a user would have any reason to apply them to the return value of one of our existing functions, so it seems unlikely that not having them will break any existing code.  But that is just a guess--one never knows what funky things users will do out there in the wild.

At this point, my "philosophy" (for lack of a better word) is to keep the API for this new object to the bare minimum.  The less we promise now, the less we have to worry about maintaining in the future.  The question is: how much of the `tuple` API do we need so that existing code doesn't break?   The features that I'm *sure* we need are:

* Indexing and unpacking, e.g. `z, p = function(a, b, c)`, `z = function(a, b, c)[0]`, etc.  The length of the sequence is fixed to be the length of the tuple currently returned by the function, even if new outputs are added to the object in later.
* For functions that currently use a `namedtuple`, the new object must continue to support accessing the values via attributes on the object.
* To add new outputs to an existing function, the new results are stored as attributes on the output object.  The new outputs are not in the tuple-like API--they are not available via indexing or unpacking.
* The object must be picklable. (An existing function might be the target for multiprocessing, and multiprocessing uses pickle to transfer Python objects between processes.)

Are there any other properties that we are *sure* we must support?

**Comment by larsoner** [comment]
> I find it hard to imagine a case where a user would have any reason to apply them to the return value of one of our existing functions, so it seems unlikely that not having them will break any existing code. But that is just a guess--one never knows what funky things users will do out there in the wild.

I'm pretty sure I've done stuff like `function_output + (some, other, values, in, another, tuple)` (not in `scipy.stats` but elsewhere) -- after all, if a function returns a tuple, it should be safe to assume it will continue to do so. But this is just my example -- I think it's safe to assume that if we don't support all tuple-like behavior, we *will* break some existing working code. So I think there is another question that we should consider as well: how many less common / rare cases are we willing to break (or at least risk breaking) by only supporting some minimal set of functions from the total set provided by `tuple`/`namedtuple` returns?

Again I think it's okay to do this if we state clearly in the release notes that people relying on things actually being a tuple should cast with `tuple(output)`. I think we'll want to do this unless we do manage to successfully subclass `tuple`, so I think the minimal set you enumerate above is fine.

Personally also I think it's worth it to add the `abc` subclass because it adds very little code but immediately makes clear (to maintainers, even) what methods they can expect this subclass to provide, adds functionality, etc. but I can live without it.

**Comment by larsoner** [comment]
> (But I'm not Raymond Hettinger--maybe my Python wizarding skills just aren't strong enough.) 

Mine probably aren't either, but I wouldn't mind trying :) @WarrenWeckesser do you have an implementation that I can work from that does subclass `tuple` or `namedtuple`? I looked back at your PRs and it looks like maybe you push-forced over those (?). Should I just work from https://github.com/scipy/scipy/issues/3665#issuecomment-636600527 ?

**Comment by WarrenWeckesser** [comment]
Adding the abc `Sequence` says "I implement `__len__` and `__getitem__`", and adds the methods `count()` and `index()` automatically.  That's fine with me.

Another thing to think about: I expect that eventually (i.e. in a few years), we're going to want to deprecate the tuple-like behavior.  (New SciPy developer in 2024: "Why do these functions have these weird return values?"  Grey beard SciPy developer: "Well, kiddo, that goes back to pre-COVID-19 days...") I think when that time comes, we would just deprecate the dunder methods `__len__` and `__getitem__`.

**Comment by larsoner** [comment]
I took at stab at adapting the `namedtuple` CPython code for our purposes, and at least all the tests here still pass:

https://github.com/WarrenWeckesser/scipy/pull/1

It's a lot more code but it preserves `tuple` checking (and maybe behavior) at least.

**Comment by WarrenWeckesser** [comment]
> ... do you have an implementation that I can work from that does subclass tuple or namedtuple? 

I put a version of a function that generates a subclass of a `namedtuple` in a gist: https://gist.github.com/WarrenWeckesser/5230e3fc24c83f64b54bdcc212354e64

Example:
```
In [2]: Result = make_namedtuple_bunch1('Result', ['x', 'y'], ['w', 'beta'])

In [3]: result = Result(x=1, y=2, w=99, beta=0.5)

In [4]: isinstance(result, tuple)
Out[4]: True

In [5]: x0, y0 = result

In [6]: result.beta
Out[6]: 0.5
```
Because of how the class is generated dynamically, instances such as `result` are not picklable.

I might have tried too hard to subclass `namedtuple`.  I also tried to keep both the class creation and object instantiation as just one call each.  Perhaps a better approach would be to skip the use of `namedtuple`, and create a subclass of `tuple` at the top-level of the module, with a different API for defining the tuple's attribute names and the names of the extra fields.

**Comment by larsoner** [comment]
> Perhaps a better approach would be to skip the use of namedtuple, and create a subclass of tuple at the top-level of the module, with a different API for defining the tuple's attribute names and the names of the extra fields.

This is basically what I did by taking the CPython `namedtuple` code and adapting it. See if you think it makes sense!

**Comment by WarrenWeckesser** [comment]
@larsoner, I just pushed an update after merging your pull request.  The changes are summarized in the commit message.

**Comment by larsoner** [comment]
Looks good to me!

**Comment by WarrenWeckesser** [comment]
OK, I squashed the commits and updated the status of the PR to "ready for review".

File: scipy/_lib/_bunch.py
Comment (mdhaber) [inline scipy/_lib/_bunch.py:40]: If `field_names` is a single string, `_validate_names` fails with a message that's not super helpful.
```
>>> Result = make_tuple_bunch('Result', 'ab')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\matth\Desktop\scipy\scipy\_lib\_bunch.py", line 40, in make_tuple_bunch
    _validate_names(typename, field_names, extra_field_names)
  File "C:\Users\matth\Desktop\scipy\scipy\_lib\_bunch.py", line 12, in _validate_names
    for name in [typename] + field_names + extra_field_names:
TypeError: can only concatenate list (not "str") to list
```
Is entering a single string for `field_names` something we want to protect against?

File: scipy/_lib/tests/test_bunch.py
Comment (mdhaber) [inline scipy/_lib/tests/test_bunch.py:22]: Is it worth adding obvious tests like:
```
assert_equal(self.result, (1, 2, 3))
assert_equal(hash(self.result), hash((1, 2, 3)))
```

**Review by mdhaber — COMMENTED** [review]
One question I have had is why we are careful to ensure that the extra fields are not mutable and that none can be added? `OptimizeResult`s don't have this restriction.

File: scipy/_lib/_bunch.py
Comment (WarrenWeckesser) [reply scipy/_lib/_bunch.py:119]: I'm not too worried about it.  The API for creating these things is strictly for SciPy developers, and as long as some sort of error occurs, the developer will know that something needs to fixed.    A user will never encounter this situation.

I was going to say "a developer should know from reading the docstring of make_tuple_bunch that the argument must be a list", but then I recalled that there was no docstring. 😮   So I added one in a new commit.

Having said that, if others think we need more validation here, I'm not against adding it.  Friendlier error messages are good for developers, too. :)

Note that the `namedtuple` code from which this is derived has code that allows the argument to be either a list of names or a single string of comma-separated names.  I'm not interested in adding that here; it seems like more code to be maintained and tested that provides no real benefit (except maybe typing a few less characters).

**Comment by WarrenWeckesser** [comment]
@mdhaber wrote
> One question I have had is why we are careful to ensure that the extra fields are not mutable and that none can be added? OptimizeResults don't have this restriction.

I had `namedtuple` on the brain, and so was thinking this object should have an API that is locked down.  You can't add attributes to a tuple or namedtuple.  For now, I think I'd like to keep it locked down.  By doing so, we aren't removing an existing feature, since this is only a replacement for objects that are tuples or namedtuples. If it turns out that there is demand from users to assign new attributes to one of these object, or to allow the extra attributes to be modified, we can enable that later.

File: scipy/_lib/tests/test_bunch.py
Comment (WarrenWeckesser) [reply scipy/_lib/tests/test_bunch.py:33]: Done.  I pushed a commit that addresses this comment and your other comments about testing other tuple-like properties.

**Review by mdhaber — APPROVED** [review]
Docstring and new tests look good to me. There are no tests where `extra_field_names` is not provided or empty, but I don't see a reason that it wouldn't work without them.

I'm curious about the naming conventions in `scipy._lib`. When should the function names be prefixed with an underscore - when they are not intended to be imported anywhere, perhaps? And since `make_tuple_bunch` is intended to be imported by other SciPy code, it is not prefixed with an underscore?

**Comment by WarrenWeckesser** [comment]
The test `test_extra_field_names_is_optional` tests calling `make_tuple_bunch` without `extra_field_names`.

I don't know of any explicitly stated rules about when function names in `_lib` should be prefixed with an underscore.  My thinking <strike>is</strike> was that, since the only way to access `make_tuple_bunch` is via `_lib`, and `_lib` is private, `make_tuple_bunch` is also private.  In fact, the underscore prefix in `_bunch` isn't strictly necessary.  But I don't know if that is a common mental model for how we should indicate objects are private.   And since you brought it up, and since the name might end up in a namespace that could appear to be public (e.g. `scipy.stats.stats`), I think I'll exercise an overabundance of caution and add an underscore prefix to `make_tuple_bunch`.

**Comment by mdhaber** [comment]
> The test `test_extra_field_names_is_optional`..

Oops, that's exactly what I was looking for. 

@larsoner Everything still look good to you? 
@WarrenWeckesser ready to merge this after prefix `_` is added?

**Comment by WarrenWeckesser** [comment]
@mdhaber, yes, I think this is ready.

The one test failure of the latest update is the recurring 60 minute timeout of the Azure Windows Python36-32bit-full test.

**Review by larsoner — APPROVED** [review]

**Comment by mdhaber** [comment]
Thanks @WarrenWeckesser @larsoner!

**Comment by WarrenWeckesser** [comment]
Thanks Eric and Matt!  This will make it much easier for us to add new features to existing functions.  Expect a flurry of PRs soon...

**Comment by WarrenWeckesser** [comment]
A case where the new bunch-like object doesn't help is when the function to be enhanced currently returns a single value, not a tuple.  Among the statistical tests in `stats`, [`binom_test`](http://scipy.github.io/devdocs/generated/scipy.stats.binom_test.html) appears to be the only case where this occurs.  We want to add a confidence interval to the return value of this function.  For this one case, perhaps it won't be too disruptive to simply deprecate `binom_test` and add `binomial_test`.

**Comment by rgommers** [comment]
Can you doc-deprecate it instead? A real deprecation in the absence of bugs or other major issues seems a bit gratuitous.

**Comment by WarrenWeckesser** [comment]
@rgommers, yes, doc-deprecation would be fine.

**Pull request merged**
